File: leetcode/js/intersectionOfTwoArrays.js
// Given two arrays, write a function to compute their intersection.

/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var intersect = function(nums1, nums2) {
  let hash = {};
  let arr1 = nums1.length > nums2.length ? nums1 : nums2;
  let arr2 = nums1.length > nums2.length ? nums2 : nums1;

  for (let i = 0; i < arr1.length; i++) {
    hash[arr1[i]] = hash[arr1[i]] || 0;
    hash[arr1[i]]++;
  }
  let res = [];
  let count = arr1.length;

  for (let i = 0; i < arr2.length && count !== 0; i++) {
    if (hash[arr2[i]] > 0) {
      hash[arr2[i]]--;
      res.push(arr2[i]);
      count--;
    }
  }
  return res;
};

File: leetcode/ruby/intersectionOfTwoArrays.rb
# Given two arrays, write a function to compute their intersection.

# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Integer[]}
def intersect(nums1, nums2)
    hash = Hash.new {|h, k| h[k] = 0 }
    nums1.each do |el|
        hash[el] += 1
    end
    output = []
    nums2.each do |el|
        if hash[el] > 0
            output << el
            hash[el] -= 1
        end
    end
    output
end